ui/lib/core/addon/components/bar-chart.js

```diff
@@ -0,0 +1,396 @@
+// eslint-disable-next-line no-unused-vars
+import { transition } from 'd3-transition';
+
+// SIZING CONSTANTS
+const CHART_MARGIN = { top: 10, left: 137 }; // makes space for y-axis legend
+const CHAR_LIMIT = 18; // character count limit (for label truncating)
+const LINE_HEIGHT = 24; // each bar w/ padding is 24 pixels thick
+
+// COLOR THEME:
+const BAR_COLOR_DEFAULT = ['#BFD4FF', '#8AB1FF'];
+const BAR_COLOR_HOVER = ['#1563FF', '#0F4FD1'];
+const BACKGROUND_BAR_COLOR = '#EBEEF2';
+const TOOLTIP_BACKGROUND = '#525761';
+
+class BarChart extends Component {
+  dataset = [
+    {
+      namespace_id: 'root',
+      namespace_path: 'root',
+      counts: {
+        distinct_entities: 268,
+        non_entity_tokens: 985,
+        clients: 1253,
+      },
+    },
+    {
+      namespace_id: 'O0i4m',
+      namespace_path: 'top-namespace',
+      counts: {
+        distinct_entities: 648,
+        non_entity_tokens: 220,
+        clients: 868,
+      },
+    },
+    {
+      namespace_id: '1oihz',
+      namespace_path: 'anotherNamespace',
+      counts: {
+        distinct_entities: 547,
+        non_entity_tokens: 337,
+        clients: 884,
+      },
+    },
+    {
+      namespace_id: '1oihz',
+      namespace_path: 'someOtherNamespaceawgagawegawgawgawgaweg',
+      counts: {
+        distinct_entities: 807,
+        non_entity_tokens: 234,
+        clients: 1041,
+      },
+    },
+  ];
+
+  get labelKey() {
+    return this.args.labelKey || 'label';
+  }
+
+  get getButtonText() {
+    return this.args.getButtonText || 'Export data';
+  }
+
+  get mapLegend() {
+    return this.args.mapLegend || null;
+  }
+
+  get chartKeys() {
+    return this.args.chartKeys || null;
+  }
+
+  // TODO: take in arguments passed to component
+  flattenedData() {
+    return this.dataset.map(d => {
+      return {
+        label: d['namespace_path'],
+        non_entity_tokens: d['counts']['non_entity_tokens'],
+        distinct_entities: d['counts']['distinct_entities'],
+        total: d['counts']['clients'],
+      };
+    });
+  }
+
+  // TODO: separate into function for specifically creating tooltip text
+  totalCount = this.dataset.reduce((prevValue, currValue) => prevValue + currValue.counts.clients, 0);
+
+  @action
+  renderBarChart(element) {
+    let totalCount = this.totalCount;
+    let handleClick = this.args.onClick;
+    let labelKey = this.labelKey;
+    let dataset = this.flattenedData();
+    let elementId = guidFor(element);
+
+    let stackFunction = stack().keys(this.mapLegend.map(l => l.key));
+    // creates an array of data for each map legend key
+    // each array contains coordinates for each data bar
+    let stackedData = stackFunction(this.flattenedData());
+
+    // creates and appends tooltip
+    let container = select('.bar-chart-container');
+    container
+      .append('div')
+      .attr('class', 'chart-tooltip')
+      .attr('style', 'position: fixed; opacity: 0;')
```

Comment:
There are quite a few styles defined here.  Can this be moved to CSS?

Reply:
@randallmorey - I wondered the same thing. I received advice at the beginning of building this component to keep all chart-related styling in d3...but I haven't researched enough to know if there are benefits of that, or if I should convert it all to CSS. I'd prefer to have all styling happen in one place.